### ISSUE TITLE
Extended block snapshot functionality

### ISSUE THREAD

**Issue by kashike** [SpongePowered/SpongeAPI#1018]
We currently have [a few methods](https://github.com/SpongePowered/SpongeAPI/blob/master/src/main/java/org/spongepowered/api/world/extent/Extent.java#L150-L231) for working with block snapshots, but they all deal with a _single_ block location/snapshot.

It would be nice to be able to create a "bulk" snapshot, that can contain many block locations/snapshots:
```java
BulkBlockSnapshot createSnapshot(Vector3i position, Vector3i size);

BulkBlockSnapshot createSnapshot(int x, int y, int z, int xSize, int ySize, int zSize);

boolean restoreSnapshot(BulkBlockSnapshot snapshot, boolean notifyNeighbors);

boolean restoreSnapshot(Vector3i position, BulkBlockSnapshot snapshot, boolean notifyNeighbors);

boolean restoreSnapshot(int x, int y, int z, BulkBlockSnapshot snapshot, force, boolean notifyNeighbors);
```
I've removed the `force` boolean from the above `restoreSnapshot` methods, since it does not seem applicable in bulk snapshots.

In addition to the above, the following would be nice, helpful, additions:
- ability to _cut_ (not just copy, as snapshots do now) when creating a snapshot
- ability to exclude __air__ blocks when creating a snapshot (as seen in WorldEdit)

**Comment by ryantheleach** [comment]
I'm not sure I  understand the point. Wouldn't it almost always be better to use a Collection<Transaction<BlockSnapshot>> or a Transaction<Collection<BlockSnapshot>> depending on the atomicity required?

Or are you thinking things like PopulatorObjects?

**Comment by kashike** [comment]
@ryantheleach Performance is a factor. Think of a region with thousands of blocks that I want to snapshot - I can either have __X__ `BlockSnapshot`s or __one__ `BulkBlockSnapshot`.

**Comment by simon816** [comment]
Not sure but I think this is close to what an `Extent` is supposed to be, except extent is mutable, so perhaps an immutable extent?

**Comment by kashike** [comment]
`BulkBlockSnapshot` should act the same as a `BlockSnapshot`, but be bulk.

**Comment by DDoS** [comment]
Part of the current view API is support for creating block volumes as copies of subregions of existing block volumes, with the option for immutable copies. There isn't a copy back function yet, that's part of the work view API I haven't started. What would BulkBlockSnapshot provide missing in ImmutableBlockVolume?

**Comment by kashike** [comment]
@DDoS Assuming `BlockVolume` does what I describe in the description (a `BlockSnapshot`, but with many blocks instead of a single block), it'll work. The question is how to create one (similar to block snapshots/how it is described in the description), and if you can exclude air blocks and cut instead of copy.

**Comment by DDoS** [comment]
There isn't a cut functionality, but that could be implemented using the planned view worker API. Similarly for filtering when copying/pasting.

**Comment by kashike** [comment]
and how do you create and restore a `BlockVolume`?

**Comment by DDoS** [comment]
Right now you create one by getting a sub-view with the min/max coordinates, then creating a copy of it. 

Restoring isn't available yet, but can be implemented with a bi-function as a merge operation. From the target view (the one being restored to) you obtain a sub-view of the same min/max (or just reuse the one used to get the copy) and get a synchronous worker view from it.

You call the work method with the bi-function and blocks from the source and target view will be passed to it. It returns the block to be set in the target view.

Alternatively you could use the a regular function if no merge is needed. Same process.

**Comment by DDoS** [comment]
Also here are some examples for using the current view API: https://github.com/SpongePowered/Cookbook/blob/master/Plugin/TestSuite/src/main/java/org/spongepowered/cookbook/plugin/BlockBufferTest.java

**Comment by ryantheleach** [comment]
Do ImmutableBlockVolumes have to be a cuboid? That's the only real thing that I can see that would be the difference between a possibly sparse bulkblocksnapshot.

**Comment by DDoS** [comment]
Not really, but it works better that way. You could implement hasBlock() to have a sparse configuration.
